<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Lab: Differential Expression via RNA-Seq Analysis</title>
    <meta charset="utf-8" />
    <meta name="author" content="Sanzhen Liu" />
    <script src="libs/header-attrs-2.14/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Lab: Differential Expression via RNA-Seq Analysis
]
.subtitle[
## Genomic Technologies Workshop 2022 (PLPTH885)
]
.author[
### Sanzhen Liu
]
.date[
### 6/8/2022
]

---

### Outline
- Differential expression test using DESeq2

- Result visualization

- GO enrichment test

### Course webpage
[RNA-seq DE analysis](https://github.com/liu3zhenlab/teaching/tree/master/RNA-Seq-Workshop/2022)


### Lab co-teacher
Dr. Guifang Lin

---
### OnDemand at Beocat
[ondemand](https://ondemand.beocat.ksu.edu/)

login with eID

&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/Ondemand.png" width=750 height=75&gt;
&lt;/div&gt;

---
### Select RStudio
&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/app_menu.png" width=200 height=500&gt;
&lt;/div&gt;

---
### Request resources
&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/Rstudio_demand.png" width=600 height=500&gt;
&lt;/div&gt;

---
### Connect to RStudio
&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/Rstudio_job.png" width=750 height=250&gt;
&lt;/div&gt;

---
### Rstudio interface

&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/Rstudio.png" width=650 height=500&gt;
&lt;/div&gt;

---
### Package installation

```r
if (!require("BiocManager", quietly=T))
  install.packages("BiocManager")

if (!require("DESeq2", quietly=T))
  BiocManager::install("DESeq2") # DESeq2

if (!require("goseq", quietly=T))
  BiocManager::install("goseq") # GOSeq

if (!require("GO.db", quietly=T))
  BiocManager::install("GO.db", force=T) # GO.db
```

---
### preload modules

```r
pls=paste0(data_url, "/utils/load.R")
source(pls)
```

- panel.cor2
- rnaseq.pca
- normalization

[codes](https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/utils/load.R)

---
### Read expression data (Read counts per gene)


```r
rc &lt;- paste0(data_url, "/data/rc.txt")
grc &lt;- read.delim(rc)
nrow(grc)  # the number of rows/lines
```

```
[1] 22697
```

|Gene             | ExonSize| ck_rep1| ck_rep2| ck_rep3| trt_rep1| trt_rep2| trt_rep3|
|:----------------|--------:|-------:|-------:|-------:|--------:|--------:|--------:|
|AC147602.5_FG004 |      483|    5480|    6075|    5934|     3370|     5784|     6432|



|      |Gene          | ExonSize| ck_rep1| ck_rep2| ck_rep3| trt_rep1| trt_rep2| trt_rep3|
|:-----|:-------------|--------:|-------:|-------:|-------:|--------:|--------:|--------:|
|22697 |GRMZM5G899985 |      615|     267|     327|     348|       83|      342|      403|

---
### RPKM normalization

```r
datacols &lt;- c("ck_rep1", "ck_rep2", "ck_rep3",
              "trt_rep1", "trt_rep2", "trt_rep3")
lib.sizes &lt;- colSums(grc[, datacols]) # library sizes
exon.sizes &lt;- grc$ExonSize
grcn &lt;- normalization(counts = grc,
                      normcolname = datacols,
                      libsize = lib.sizes,
                      exonsize = exon.sizes,
                      methods = "RPKM")
```


|Gene             | ExonSize| ck_rep1| ck_rep2| ck_rep3| trt_rep1| trt_rep2| trt_rep3| ck_rep1.RPKM| ck_rep2.RPKM| ck_rep3.RPKM| trt_rep1.RPKM| trt_rep2.RPKM| trt_rep3.RPKM|
|:----------------|--------:|-------:|-------:|-------:|--------:|--------:|--------:|------------:|------------:|------------:|-------------:|-------------:|-------------:|
|AC147602.5_FG004 |      483|    5480|    6075|    5934|     3370|     5784|     6432|      854.123|      895.760|      904.373|       567.493|       915.326|       916.971|
|AC148152.3_FG005 |     1422|     187|     295|     377|      169|      158|      563|        9.900|       14.775|       19.516|         9.666|         8.493|        27.262|

---
### data organization for DESeq2
- count information

```r
geneid &lt;- grc$Gene
in.data &lt;- as.matrix(grc[, 3:8])
```


| ck_rep1| ck_rep2| ck_rep3| trt_rep1| trt_rep2| trt_rep3|
|-------:|-------:|-------:|--------:|--------:|--------:|
|    5480|    6075|    5934|     3370|     5784|     6432|

---
### sample names and grouping information (treatment)

```r
sample.ids &lt;- colnames(in.data)
treatment &lt;- c("ck", "ck", "ck", "trt", "trt", "trt")
sample.info &lt;- data.frame(row.names=sample.ids, trt=treatment)
```

|         |trt |
|:--------|:---|
|ck_rep1  |ck  |
|ck_rep2  |ck  |
|ck_rep3  |ck  |
|trt_rep1 |trt |
|trt_rep2 |trt |
|trt_rep3 |trt |

---
### Differential expression test


```r
dds &lt;- DESeqDataSetFromMatrix(countData=in.data,
                              colData=sample.info,
                              formula(~trt))
dds &lt;- DESeq(dds)
```

---
### DE output

```r
res &lt;- results(object = dds)
res &lt;- data.frame(res)
res$Gene &lt;- geneid
res &lt;- res[,c("Gene","baseMean","log2FoldChange","pvalue","padj")]
nrow(res)
```

```
[1] 22697
```

---
### DE + normalized data

```r
### Merge the normalized result with the DE result
out &lt;- merge(grcn, res, by = "Gene")
out &lt;- data.frame(out)
```


|Gene             | ExonSize| ck_rep1| ck_rep2| ck_rep3| trt_rep1| trt_rep2|
|:----------------|--------:|-------:|-------:|-------:|--------:|--------:|
|AC147602.5_FG004 |      483|    5480|    6075|    5934|     3370|     5784|
|AC148152.3_FG005 |     1422|     187|     295|     377|      169|      158|



| trt_rep3| ck_rep1.RPKM| ck_rep2.RPKM| ck_rep3.RPKM| trt_rep1.RPKM|
|--------:|------------:|------------:|------------:|-------------:|
|     6432|      854.123|      895.760|      904.373|       567.493|
|      563|        9.900|       14.775|       19.516|         9.666|



| trt_rep2.RPKM| trt_rep3.RPKM|  baseMean| log2FoldChange|    pvalue|      padj|
|-------------:|-------------:|---------:|--------------:|---------:|---------:|
|       915.326|       916.971| 5441.6579|     -0.1490702| 0.4642180| 0.8049379|
|         8.493|        27.262|  285.5493|      0.0431574| 0.9258171| 0.9845702|

---
#### significant gene sets at different FDRs

```r
sum(!is.na(out$padj) &amp; out$padj &lt; 0.05)
```

```
[1] 1261
```

**problem**  
Please revise the code to calculate the number of significant genes with the FDR smaller than 10% and 15%?

---
### significantly DEG

```r
sig &lt;- out[!is.na(out$padj) &amp; out$padj &lt; 0.05, ]
```

---
### p-value histogram

```r
pvals &lt;- out$pvalue
hist(pvals, main="Histogram",xlab="p-values",ylab="Number of genes")
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-18-1.png" style="display: block; margin: auto;" /&gt;

**problem**  
Please modify the plot code to change the figure title to "DE"

---
### scatter plot - raw counts

```r
x.raw &lt;- out$ck_rep1
y.raw &lt;- out$ck_rep2
drange &lt;- range(x.raw, y.raw, finite=T)
# plot a scatter plot between two samples
plot(x.raw, y.raw, main = "Raw counts",
     xlab="control rep1", ylab="control rep2",
     xlim = drange, ylim = drange)
abline(a = 0, b = 1, col = "orange")
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-19-1.png" style="display: block; margin: auto;" /&gt;

---
### scatter plot - RPKM

```r
x &lt;- log(out$ck_rep1.RPKM)
y &lt;- log(out$ck_rep2.RPKM)
drange &lt;- range(x, y, finite=T)
plot(x, y, main = "log of RPKM",
     xlab = "control rep1", ylab = "control rep2",
     xlim = drange, ylim = drange)
abline(a = 0, b = 1, col = "orange")
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-20-1.png" style="display: block; margin: auto;" /&gt;

---
### pair-wise scatter plots

```r
logrpkm &lt;- log(out[, 9:14])
pairs(logrpkm, lower.panel=panel.smooth, upper.panel=panel.cor2)
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-21-1.png" style="display: block; margin: auto;" /&gt;

---
### Principal Component Analysis (PCA)
&lt;div align="center"&gt;
&lt;img src="https://raw.githubusercontent.com/liu3zhenlab/teaching/master/RNA-Seq-Workshop/2022/images/pca.slide.png" width=750 height=400&gt;
&lt;/div&gt;

---
### function / module
You can write your own function:  
fun_name &lt;- function (…) {
  ...
}

```r
gpa_improve &lt;- function(gpa, rate) {
### gpa: a numeric vector for GPAs
### rate: percentage for the improvement
	new.gpa &lt;- gpa * (1 + rate)
	new.gpa[new.gpa &gt; 4] &lt;- 4	
	return(new.gpa)
}

### running the function
our.gpa &lt;- c(3.8, 3.3, 2.8, 3.1)
gpa_improve(our.gpa, 0.1)
```

```
[1] 4.00 3.63 3.08 3.41
```

```r
gpa_improve(our.gpa, 0.2)
```

```
[1] 4.00 3.96 3.36 3.72
```

---
### PCA function

**principal component analysis and ploting**

```r
rnaseq.pca &lt;- function(norm.data,  
  norm.feature="RPKM",  
	group.feature,  
	title="",  
  shape.code=NULL,  
	mean.cutoff=0.1,  
	colors=NULL,
	scaling=T, ...) {  
	…  
}
```

---
### PCA plotting

```r
rnaseq.pca(out, norm.feature="RPKM", group.feature=c("ck", "trt"),
           cex=3, shape.code=c(5, 16), mean.cutoff=0.1, scaling = T,
           title="PCA-all genes", colors=c("dark green","orange"))
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-24-1.png" style="display: block; margin: auto;" /&gt;

---
### MA plot

```r
aval &lt;- log(out$baseMean)
mval &lt;- out$log2FoldChange
sig.set &lt;- which(out$padj &lt; 0.05)
plot(aval, mval, main="MA plot",
     xlab="log(mean of exp)", ylab="log2FC",
     ylim=c(-4, 6), pch=19, cex=0.3, col="blue")
# add points and a horizontal line
points(aval[sig.set], mval[sig.set], pch=19,cex=0.3,col="red")
abline(h = 0, cex = 1, col = "dark green")
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-25-1.png" style="display: block; margin: auto;" /&gt;

---
### Volcano plot

```r
log2val &lt;- out$log2FoldChange  # log2 fold change
mlogP &lt;- -log10(out$pvalue) # transformed p-values
### plot
plot(log2val, mlogP, main="Volcano plot",
     xlab="log2FC", ylab="-log10(p-value)",
     pch=19, cex=0.4,col="grey", xlim=c(-6,6), ylim=c(0,16))
### highlight
up &lt;- which(out$padj&lt;0.05 &amp; log2val &gt; 0) # up
down &lt;- which(out$padj&lt;0.05 &amp; log2val &lt; 0)  # down
points(log2val[up], mlogP[up], pch=19,cex=0.4,col="brown")
points(log2val[down], mlogP[down],pch=19,cex=0.4,col="blue")
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-26-1.png" style="display: block; margin: auto;" /&gt;

---
### Gene ontology (GO) enrichment analysis
- a gene and GO association table

- a list of all genes

- a list of significant genes

- mean or total gene read counts per gene (**optional**)

---
### GOSeq (I)

```r
gdbf=paste0(data_url, "/data/go.txt")
godb &lt;- read.delim(gdbf)
geneid &lt;- as.character(out$Gene)  # gene vector
# a vector to indicate if the gene is DE (0 or 1)
de.vector &lt;- as.integer(!is.na(out$padj) &amp; out$padj &lt; 0.05)
names(de.vector) &lt;- geneid
countbias &lt;- out$baseMean # total raw reads per gene
# bias fitting
pwf.counts &lt;- nullp(DEgenes=de.vector, bias.data=countbias)
```

&lt;img src="RNASeq-DE_files/figure-html/unnamed-chunk-27-1.png" style="display: block; margin: auto;" /&gt;

---
### GOSeq (II)

```r
go &lt;- goseq(pwf=pwf.counts, gene2cat=godb, method="Sampling",
            repcnt=1000, use_genes_without_cat=F)  # GO enrichment
```


|                         |208                    |
|:------------------------|:----------------------|
|category                 |GO:0004175             |
|over_represented_pvalue  |0.000999001            |
|under_represented_pvalue |1                      |
|numDEInCat               |16                     |
|numInCat                 |31                     |
|term                     |endopeptidase activity |
|ontology                 |MF                     |


```r
example.go &lt;- GOTERM[["GO:0004175"]]  # GO information
Definition(example.go)  # return GO definition
```

```
[1] "Catalysis of the hydrolysis of internal, alpha-peptide bonds in a polypeptide chain."
```

---
## Summary of the analyzing procedure

1. Read counts per gene

2. DE analysis based on the experimental design

3. Examine results (p-value distribution, number of significant genes)

4. Gene Ontology enrichment test

[DESeq2 tutorial](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

---
### Contact information
Sanzhen Liu  
Plant Pathology  
4729 Throckmorton Plant Sciences Center  
Manhattan, KS 66506-5502  
phone: 785-532-1379  

[liu3zhen@ksu.edu](liu3zhen@ksu.edu)  
twitter: [liu3zhen](https://twitter.com/liu3zhen?lang=en)

### Bioinformatics Applications
PLPTH813, Spring 2022
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
